webextension/privileged/ui/CloudDownloadsView.jsm

```diff
@@ -0,0 +1,760 @@
+   * menuItem and moveDownloadSubMenu (if multiple providers) inside DownloadsContextMenu
+   * b) an event listener to listen to contextmenu event on downloadsListBox
+   * MenuItems are updated with label attributes before displaying Move Download option inside
+   * _displayMoveToCloudContextMenuItem
+   */
+  async registerContextMenu(browserWindow) {
+    try {
+      if (!browserWindow || !browserWindow.document) {
+        return;
+      }
+
+      // There is no other API that awaits the result of CloudStorageInternal.init()
+      // Invoke getDownloadFolder on CloudStorage API to ensure API is initialized
+      // before asking CloudStorage API for storage providers info
+      await CloudStorage.getDownloadFolder();
+      const providers = await CloudStorage.getStorageProviders();
+      this.providers = await this._filterProvidersMap(providers);
+
+      // Continue only if cloud providers exists on user device
+      if (this.providers.size === 0) {
+        return;
+      }
+
+      const aPopupMenu = browserWindow.document.getElementById("downloadsContextMenu");
+      if (!aPopupMenu) {
+        return;
+      }
+
+      // Get menuItem 'Copy Location', to insert skeleton 'Move download' menuitem and a menu separator after it
+      const menuItem = aPopupMenu.getElementsByAttribute("command", "downloadsCmd_copyLocation")[0];
+      if (!menuItem) {
+        return;
+      }
+
+      const fragment = browserWindow.document.createDocumentFragment();
+      const separator = browserWindow.document.createElement("menuseparator");
+      separator.setAttribute("id", "moveDownloadSeparator");
+      fragment.appendChild(separator);
+
+      if (this.providers.size > 1) {
+        const moveDownloadMenu = browserWindow.document.createElement("menu");
+        moveDownloadMenu.setAttribute("id", "moveDownload");
+        moveDownloadMenu.setAttribute("hidden", "true");
+        const moveDownloadMenuPopup = browserWindow.document.createElement("menupopup");
+        moveDownloadMenuPopup.setAttribute("id", "moveDownloadSubMenu");
+        moveDownloadMenuPopup.setAttribute("hidden", "true");
+
+        this.providers.forEach((value, key) => {
+          const moveDownloadItem = browserWindow.document.createElement("menuitem");
+          moveDownloadMenuPopup.appendChild(moveDownloadItem);
+        });
+
+        moveDownloadMenu.appendChild(moveDownloadMenuPopup);
+        fragment.appendChild(moveDownloadMenu);
+      } else {
+        // Add skeleton Move to menu item in context menu
+        const moveDownloadItem = browserWindow.document.createElement("menuitem");
+        moveDownloadItem.setAttribute("id", "moveDownload");
+        moveDownloadItem.setAttribute("hidden", "true");
+        fragment.appendChild(moveDownloadItem);
+      }
+      aPopupMenu.insertBefore(fragment, menuItem.nextSibling);
+      aPopupMenu.addEventListener("command", this);
+      aPopupMenu.addEventListener("popuphidden", this);
```

Comment:
You never remove this listener, which will cause leaks.

But also, I don't see any code handling the event. From your earlier comment:

> @gijsk would like to use this event to capture when downloadContextMenu is hidden, to hide 'moveDownload' menuitem in downloadContextMenu. This is needed to avoid showing moveDownload option in context menu from Library -> Downloads -> Right click download item.

I'm not really sure I understand. Why can't we show the item in the list of downloads in the library? Shouldn't it show up there, too? In any case, I think the best way would be to replace the `contextmenu` listener you have with a `popupshowing` listener that you attach to the context menu popup where you're adding items. Then that will fire no matter where that context menu opens, and when it fires you can detect whether the `triggerNode` for the popup is in the downloads list or somewhere else, and hide items appropriately (and bail out early, I guess, avoiding the more complicated logic in `_displayMoveToCloudContextMenuItem`).

Reply:
Library -> Downloads as of now is out of scope and nice to have for this study. Updated PR to use popupshowing and bail out early for triggerNode outside of downloadsListBox